base/associative.jl

```diff
@@ -215,13 +253,45 @@ Dict{String,Float64} with 3 entries:
   "foo" => 0.0
 ```
 """
-function merge(d::Associative, others::Associative...)
+merge(d::Associative, others::Associative...) =
+    merge!(emptymergedict(d, others...), d, others...)
+
+"""
+    merge(combine, d::Associative, others::Associative...)
+
+Construct a merged collection from the given collections. If necessary, the
+types of the resulting collection will be promoted to accommodate the types of
+the merged collections. Values with the same key will be combined using the
+combiner function.
+
+```jldoctest
+julia> a = Dict("foo" => 0.0, "bar" => 42.0)
+Dict{String,Float64} with 2 entries:
+  "bar" => 42.0
+  "foo" => 0.0
+
+julia> b = Dict("baz" => 17, "bar" => 4711)
+Dict{String,Int64} with 2 entries:
+  "bar" => 4711
+  "baz" => 17
+
+julia> merge(+, a, b)
+Dict{String,Float64} with 3 entries:
+  "bar" => 4753.0
+  "baz" => 17.0
+  "foo" => 0.0
+```
+"""
+merge(combine::Function, d::Associative, others::Associative...) =
+    merge!(combine, emptymergedict(d, others...), d, others...)
+
+function emptymergedict(d::Associative, others::Associative...)
     K, V = keytype(d), valtype(d)
```

Comment:
This implementation:

```julia
promoteKV(K,V) = (K,V)
promoteKV(K,V, d, ds...) = promoteKV(promote_type(K, keytype(d)), promote_type(V, valtype(d)), ds...)
function Base.emptymergedict(d::Associative, others::Associative...)
    K, V = promoteKV(keytype(d), valtype(d), others...)
    Dict{K,V}()
end
```

results in:
```julia
julia> d1 = Dict(Int32(1) => 1, Int32(2) => 2);

julia> d2 = Dict(1 => 1.0, 2 => 2.0);

julia> @inferred Base.emptymergedict(d1, d2)
ERROR: return type Dict{Int64,Float64} does not match inferred return type Dict{_,_} where _ where _
Stacktrace:
 [1] error(::String) at ./error.jl:21
```

But I tried this:
```julia
promoteK(K) = K
promoteV(V) = V
promoteK(K, d, ds...) = promoteK(promote_type(K, keytype(d)), ds...)
promoteV(V, d, ds...) = promoteV(promote_type(V, valtype(d)), ds...)
function Base.emptymergedict(d::Associative, others::Associative...)
    K = promoteK(keytype(d), others...)
    V = promoteV(valtype(d), others...)
    Dict{K,V}()
end
```
which seems to work:

```julia
julia> d1 = Dict(Int32(1) => 1, Int32(2) => 2);

julia> d2 = Dict(1 => 1.0, 2 => 2.0);

julia> @inferred Base.emptymergedict(d1, d2)
Dict{Int64,Float64} with 0 entries
```

Should I update to that version? Should `promoteK/V` be inlined?